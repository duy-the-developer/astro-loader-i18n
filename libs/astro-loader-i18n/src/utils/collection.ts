export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>()): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    keys.add(key);
    if (value && typeof value === "object" && !Array.isArray(value)) {
      getAllUniqueKeys(value as Record<string, unknown>, keys);
    }
  });
  return keys;
}

export function convertToSingleLocale(obj: Record<string, unknown>, locales: string[], locale: string): Record<string, unknown> {
  return Object.entries(obj).reduce(
    (result, [key, value]) => {
      if (key === locale) {
        Object.assign(result, value);
      } else if (!locales.includes(key)) {
        result[key] =
          typeof value === "object" && !Array.isArray(value) ? convertToSingleLocale(value as Record<string, unknown>, locales, locale) : value;
      }
      return result;
    },
    {} as Record<string, unknown>
  );
}
