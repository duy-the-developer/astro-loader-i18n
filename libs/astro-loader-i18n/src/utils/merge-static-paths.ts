import { GetStaticPathsResult } from "astro";

export function mergeStaticPaths(...staticPaths: GetStaticPathsResult[]): GetStaticPathsResult {
  if (staticPaths.some((paths) => paths.length !== staticPaths[0].length)) {
    throw new Error("All static paths must have the same length");
  }

  return staticPaths[0].map((entry, index) => {
    return staticPaths.reduce(
      (merged, current) => ({
        params: { ...merged.params, ...current[index].params },
        props: { ...merged.props, ...current[index].props },
      }),
      entry
    );
  });
}
