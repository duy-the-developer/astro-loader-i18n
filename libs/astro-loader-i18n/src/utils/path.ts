const TRIM_SLASHES_PATTERN = /^\/|\/$/g;
const DOUBLE_POINTS_PATTERN = /\.{2}/g;
const DOUBLE_SLASH_PATTERN = /\/{2}/g;
const INDEX_COMMON_TRANSLATION_ID = "index";

export function joinPath(...paths: Array<string | number | undefined>) {
  return paths.filter(Boolean).join("/");
}

export function resolvePath(...paths: Array<string | number | undefined>) {
  return `/${joinPath(...paths)}`;
}

export const trimSlashes = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_SLASHES_PATTERN, "");
};

export const parseLocale = (path: string, locales: string[], defaultLocale: string) => {
  const locale = locales.find((locale) => path.indexOf(locale) !== -1);

  if (locale) {
    return locale;
  }

  return defaultLocale;
};

export const createTranslationId = (path: string, pathLocale: string) => {
  path = path.replace(pathLocale, "");
  path = path.replace(DOUBLE_POINTS_PATTERN, ".");
  path = path.replace(DOUBLE_SLASH_PATTERN, "/");
  path = trimSlashes(path);

  if (path === "" || path === "/") return INDEX_COMMON_TRANSLATION_ID;

  return path;
};
