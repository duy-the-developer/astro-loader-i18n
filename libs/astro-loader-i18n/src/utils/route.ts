import { resolvePath } from "./path";

type Segments = Record<string, string | undefined>;
export type SegmentTranslations = Record<string, Segments>;
type RouteSegment = {
  spread: boolean;
  param: boolean;
  value: string;
};
type RoutePattern = RouteSegment[];

/**
 * Parses a route pattern into an array of route segments.
 *
 * @param routePattern is a string that templates the a route (e.g. /[...locale]/blog/[slug])
 */
export function parseRoutePattern(routePattern: string): RoutePattern {
  const segments = routePattern
    .split("/")
    .filter((segment) => segment !== "")
    .map((segment) => {
      if (segment.startsWith("[") && segment.endsWith("]")) {
        return {
          spread: segment.startsWith("[..."),
          param: true,
          value: segment.replace("[", "").replace("...", "").replace("]", ""),
        };
      }

      return {
        spread: false,
        param: false,
        value: segment,
      };
    });

  if (segments.length === 0) {
    return [{ spread: false, param: false, value: "/" }];
  }

  return segments;
}

/**
 *
 * @param routePattern is an array consisting of route segments
 */
export function buildPath(routePattern: RoutePattern, segmentValues: Segments): string {
  return resolvePath(
    ...routePattern.map((segment) => {
      if (segment.param) {
        if (!segmentValues[segment.value] && !segment.spread) {
          throw new Error(`No segment value found for route segment ${segment.value}`);
        }

        return segmentValues[segment.value];
      }
      return `${segment.value}`;
    })
  );
}
