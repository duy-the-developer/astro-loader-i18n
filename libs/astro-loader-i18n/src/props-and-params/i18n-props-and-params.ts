import limax from "limax";
import { checkI18nLoaderCollection, i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { buildPath, parseRoutePattern, SegmentTranslations } from "../utils/route";
import { z } from "astro/zod";

type Config = {
  defaultLocale: string;
  routePattern: string;
  segmentTranslations: SegmentTranslations;
  localeParamName?: string;
  slugParamName?: string;
  titleDataKey?: string;
};

const defaultConfig = {
  localeParamName: "locale",
  slugParamName: "slug",
  titleDataKey: "title",
} as const;

function getSegmentTranslations(
  data: z.infer<typeof i18nLoaderSchema> & Record<string, string | unknown>,
  c: Omit<Required<Config>, "routePattern">
) {
  if (!c.segmentTranslations[data.locale]) throw new Error(`No slugs found for locale ${data.locale}`);

  const currentLocale = data.locale === c.defaultLocale ? undefined : data.locale;
  const segmentValues = { [c.localeParamName]: currentLocale, ...c.segmentTranslations[data.locale] };

  const slugValue = c.titleDataKey ? data[c.titleDataKey] : undefined;
  if (slugValue && typeof slugValue === "string") {
    segmentValues[c.slugParamName] = limax(slugValue);
  }
  return segmentValues;
}

export function i18nPropsAndParams<C extends Config>(collection: unknown[], config: C) {
  checkI18nLoaderCollection(collection);
  const { routePattern, ...c } = { ...defaultConfig, ...config };
  const route = parseRoutePattern(routePattern);

  route.forEach((segment, index) => {
    if (
      segment.param &&
      segment.value !== c.localeParamName &&
      index !== route.length - 1 &&
      !Object.values(c.segmentTranslations).every((translation) => translation[segment.value])
    ) {
      throw new Error(`No slugs found for route segment ${segment.value}`);
    }
  });

  return collection.map((entry) => {
    const { locale, translationId } = entry.data;
    const entryTranslations = collection.filter((e) => e.data.translationId === translationId);
    const translations = entryTranslations.reduce(
      (previous, current) => {
        return {
          ...previous,
          [current.data.locale]: buildPath(route, getSegmentTranslations(current.data, c)),
        };
      },
      {} as Record<string, string>
    );

    return {
      params: getSegmentTranslations(entry.data, c),
      props: {
        locale,
        translationId,
        translations,
      },
    };
  });
}
