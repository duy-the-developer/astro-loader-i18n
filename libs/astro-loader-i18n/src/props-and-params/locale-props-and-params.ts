import { checkI18nLoaderCollection } from "../schemas/i18n-loader-schema";
import { GetStaticPathsResult } from "astro";
import { getSegmentTranslations, SegmentTranslations } from "../utils/route";

export function localePropsAndParams(
  collection: unknown[],
  defaultLocale: string,
  segmentTranslations?: SegmentTranslations
): GetStaticPathsResult {
  checkI18nLoaderCollection(collection);

  return collection.map((entry) => {
    const segments = segmentTranslations ? getSegmentTranslations(segmentTranslations, entry.data.locale) : {};
    const { locale, translationId } = entry.data;

    let paramsLocale = locale;
    if (locale === defaultLocale) {
      paramsLocale = "";
    }

    return {
      params: { locale: paramsLocale, ...segments },
      props: { locale, translationId },
    };
  });
}
