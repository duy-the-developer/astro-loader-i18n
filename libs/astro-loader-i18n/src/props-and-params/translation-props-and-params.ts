import { GetStaticPathsResult } from "astro";
import limax from "limax";
import { checkI18nLoaderCollection } from "../schemas/i18n-loader-schema";
import { buildPath, getSegmentTranslations, parseRoutePattern, SegmentTranslations } from "../utils/route";

type Config = {
  routePattern: string;
  segmentTranslations: SegmentTranslations;
  defaultLocale: string;
  localeParamName?: string;
  slugParamName?: string;
  titleDataKey?: string;
};

const defaultConfig = {
  localeParamName: "locale",
  slugParamName: "slug",
  titleDataKey: "title",
};

export function translationPropsAndParams(collection: unknown[], config: Config): GetStaticPathsResult {
  checkI18nLoaderCollection(collection);
  const { routePattern, segmentTranslations, defaultLocale, localeParamName, slugParamName, titleDataKey } = { ...defaultConfig, ...config };
  const route = parseRoutePattern(routePattern);

  route.forEach((segment, index) => {
    if (
      segment.param &&
      segment.value !== localeParamName &&
      index !== route.length - 1 &&
      !Object.values(segmentTranslations).every((translation) => translation[segment.value])
    ) {
      throw new Error(`No slugs found for route segment ${segment.value}`);
    }
  });

  return collection.map((entry) => {
    const translationId = entry.data.translationId;
    const entryTranslations = collection.filter((entry) => entry.data.translationId === translationId);
    const translations = entryTranslations.reduce(
      (previous, current) => {
        const segmentValues = getSegmentTranslations(segmentTranslations, current.data.locale);
        segmentValues[localeParamName] = defaultLocale === current.data.locale ? "" : current.data.locale;
        const slugValue = titleDataKey ? (current.data as Record<string, string | undefined>)[titleDataKey] : undefined;
        if (slugValue) {
          segmentValues[slugParamName] = limax(slugValue);
        }
        return {
          ...previous,
          [current.data.locale]: buildPath(route, segmentValues),
        };
      },
      {} as Record<string, string>
    );

    return {
      params: {},
      props: {
        translations,
      },
    };
  });
}
