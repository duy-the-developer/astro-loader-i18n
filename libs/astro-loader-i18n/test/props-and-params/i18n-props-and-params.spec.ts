import { describe, expect, it } from "vitest";
import { i18nPropsAndParams } from "../../src/props-and-params/i18n-props-and-params";
import { SegmentTranslations } from "../../src/utils/route";

describe("i18nPropsAndParams", () => {
  it("should generate paths for a valid collection", () => {
    const routePattern = "/[...locale]/[blog]/posts/[slug]";
    const segmentTranslations: SegmentTranslations = { "de-CH": { blog: "logbuch" }, "zh-CN": { blog: "blog" } };
    const collection = [
      {
        data: { locale: "de-CH", translationId: "magic.mdx", title: "Verückte Umlaute!" },
      },
      {
        data: { locale: "zh-CN", translationId: "magic.mdx", title: "神奇的标题" },
      },
    ];
    const result = i18nPropsAndParams(collection, { routePattern, segmentTranslations, defaultLocale: "de-CH" });
    expect(result).toMatchSnapshot();
  });
});
