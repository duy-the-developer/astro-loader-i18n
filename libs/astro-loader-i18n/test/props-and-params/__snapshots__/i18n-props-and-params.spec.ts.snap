// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nPropsAndParams > should generate paths for a valid collection 1`] = `
[
  {
    "params": {
      "blog": "logbuch",
      "locale": undefined,
      "slug": "veruckte-umlaute",
    },
    "props": {
      "locale": "de-CH",
      "translationId": "magic.mdx",
      "translations": {
        "de-CH": "/logbuch/posts/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/shen2-qi2-de-biao1-ti2",
      },
    },
  },
  {
    "params": {
      "blog": "blog",
      "locale": "zh-CN",
      "slug": "shen2-qi2-de-biao1-ti2",
    },
    "props": {
      "locale": "zh-CN",
      "translationId": "magic.mdx",
      "translations": {
        "de-CH": "/logbuch/posts/veruckte-umlaute",
        "zh-CN": "/zh-CN/blog/posts/shen2-qi2-de-biao1-ti2",
      },
    },
  },
]
`;
